#### Description
The dataset is a collection of information about patients appointments in Brazil with focus on keeping to the appointments by the patients.
#### Question
Using the data, we need to determine the important factors that can help in predicting if a patient will show up for their scheduled appointment.
#### Steps
    1.Import the necessary libraries and load dataset
    2.Get basic information about the dataset.
    3.Clean the dataset
        i.Correct the datatype for the necessary columns.
        ii.Rename columns if necessary
        iii.Drop rows that are not relevant to the analysis.
    4.Analyse and use visualizations to answer the project questions.

###### Import the necessary libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

df = pd.read_csv('appointments.csv')

##### Basic Information about the dataset

In [2]:
df.head()

PatientId  AppointmentID Gender          ScheduledDay  \
0  2.987250e+13        5642903      F  2016-04-29T18:38:08Z   
1  5.589978e+14        5642503      M  2016-04-29T16:08:27Z   
2  4.262962e+12        5642549      F  2016-04-29T16:19:04Z   
3  8.679512e+11        5642828      F  2016-04-29T17:29:31Z   
4  8.841186e+12        5642494      F  2016-04-29T16:07:23Z   

         AppointmentDay  Age      Neighbourhood  Scholarship  Hipertension  \
0  2016-04-29T00:00:00Z   62    JARDIM DA PENHA            0             1   
1  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             0   
2  2016-04-29T00:00:00Z   62      MATA DA PRAIA            0             0   
3  2016-04-29T00:00:00Z    8  PONTAL DE CAMBURI            0             0   
4  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             1   

   Diabetes  Alcoholism  Handcap  SMS_received No-show  
0         0           0        0             0      No  
1         0           0        0             0      No  
2         0           0        0             0      No  
3         0           0        0             0      No  
4         1           0        0             0      No

In [3]:
df.tail()

PatientId  AppointmentID Gender          ScheduledDay  \
110522  2.572134e+12        5651768      F  2016-05-03T09:15:35Z   
110523  3.596266e+12        5650093      F  2016-05-03T07:27:33Z   
110524  1.557663e+13        5630692      F  2016-04-27T16:03:52Z   
110525  9.213493e+13        5630323      F  2016-04-27T15:09:23Z   
110526  3.775115e+14        5629448      F  2016-04-27T13:30:56Z   

              AppointmentDay  Age Neighbourhood  Scholarship  Hipertension  \
110522  2016-06-07T00:00:00Z   56   MARIA ORTIZ            0             0   
110523  2016-06-07T00:00:00Z   51   MARIA ORTIZ            0             0   
110524  2016-06-07T00:00:00Z   21   MARIA ORTIZ            0             0   
110525  2016-06-07T00:00:00Z   38   MARIA ORTIZ            0             0   
110526  2016-06-07T00:00:00Z   54   MARIA ORTIZ            0             0   

        Diabetes  Alcoholism  Handcap  SMS_received No-show  
110522         0           0        0             1      No  
110523         0           0        0             1      No  
110524         0           0        0             1      No  
110525         0           0        0             1      No  
110526         0           0        0             1      No

In [4]:
df.shape

(110527, 14)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 110527 entries, 0 to 110526
Data columns (total 14 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   PatientId       110527 non-null  float64
 1   AppointmentID   110527 non-null  int64  
 2   Gender          110527 non-null  object 
 3   ScheduledDay    110527 non-null  object 
 4   AppointmentDay  110527 non-null  object 
 5   Age             110527 non-null  int64  
 6   Neighbourhood   110527 non-null  object 
 7   Scholarship     110527 non-null  int64  
 8   Hipertension    110527 non-null  int64  
 9   Diabetes        110527 non-null  int64  
 10  Alcoholism      110527 non-null  int64  
 11  Handcap         110527 non-null  int64  
 12  SMS_received    110527 non-null  int64  
 13  No-show         110527 non-null  object 
dtypes: float64(1), int64(8), object(5)
memory usage: 11.8+ MB


From the information above, it can be seen that:

    i.there is no missing values in any of the columns;
    ii.of the 14 columns, 8 are integers, 1 is float and five are objects

#### Clean the dataset
    i.Correct the datatype for the necessary columns.
    ii.Rename columns if necessary
    iii.Check for duplicated rows
    iv.Drop rows that are not relevant to the analysis.

In [6]:
df.dtypes

PatientId         float64
AppointmentID       int64
Gender             object
ScheduledDay       object
AppointmentDay     object
Age                 int64
Neighbourhood      object
Scholarship         int64
Hipertension        int64
Diabetes            int64
Alcoholism          int64
Handcap             int64
SMS_received        int64
No-show            object
dtype: object

Observations

    i.PatientID is float. Convert to int.
    ii.ScheduleDay and Appointmentday are objects. Converts to dates.
    iii.Correct wrong column names and anc change SMS_received and No-show for consistency with others.
    iv.All the appointment days have 00:00:00 timestamp. That will be ignored. Only the          dates will be needed. Similarly, the time part of the ScheduledDay will be ignored.

In [7]:
#Convert PatientID to int
df.PatientId = df.PatientId.astype('int64')

In [8]:
#Convert ScheduleDay and AppointmentDay to datetime
df['ScheduledDay'] = pd.to_datetime(df['ScheduledDay'])
df['AppointmentDay'] = pd.to_datetime(df['AppointmentDay'])
print(df.AppointmentDay.dtypes)
print(df.ScheduledDay.dtypes)
print(df[['ScheduledDay', 'AppointmentDay']].head(1))

datetime64[ns, UTC]
datetime64[ns, UTC]
               ScheduledDay            AppointmentDay
0 2016-04-29 18:38:08+00:00 2016-04-29 00:00:00+00:00


In [9]:
#Use dt.date attribute to get rid of the time portion
df['ScheduledDay'] = pd.to_datetime(df['ScheduledDay']).dt.date
df['AppointmentDay'] = pd.to_datetime(df['AppointmentDay']).dt.date
print(df[['ScheduledDay', 'AppointmentDay']].head(2))

  ScheduledDay AppointmentDay
0   2016-04-29     2016-04-29
1   2016-04-29     2016-04-29


In [10]:
#Rename columns
df.columns

Index(['PatientId', 'AppointmentID', 'Gender', 'ScheduledDay',
       'AppointmentDay', 'Age', 'Neighbourhood', 'Scholarship', 'Hipertension',
       'Diabetes', 'Alcoholism', 'Handcap', 'SMS_received', 'No-show'],
      dtype='object')

In [11]:
df.rename(columns = {'Hipertension':'Hypertension', 'Handcap':'Handicap', 'SMS_received':'SMSReceived', 'No-show':'NoShow'}, inplace = True)

In [12]:
df.head(2)

PatientId  AppointmentID Gender ScheduledDay AppointmentDay  Age  \
0   29872499824296        5642903      F   2016-04-29     2016-04-29   62   
1  558997776694438        5642503      M   2016-04-29     2016-04-29   56   

     Neighbourhood  Scholarship  Hypertension  Diabetes  Alcoholism  Handicap  \
0  JARDIM DA PENHA            0             1         0           0         0   
1  JARDIM DA PENHA            0             0         0           0         0   

   SMSReceived NoShow  
0            0     No  
1            0     No

In [13]:
#Check for duplicated rows
df.duplicated().sum()

0

In [14]:
#Summarise the dataset
df.describe()

PatientId  AppointmentID            Age    Scholarship  \
count  1.105270e+05   1.105270e+05  110527.000000  110527.000000   
mean   1.474963e+14   5.675305e+06      37.088874       0.098266   
std    2.560949e+14   7.129575e+04      23.110205       0.297675   
min    3.921700e+04   5.030230e+06      -1.000000       0.000000   
25%    4.172614e+12   5.640286e+06      18.000000       0.000000   
50%    3.173184e+13   5.680573e+06      37.000000       0.000000   
75%    9.439172e+13   5.725524e+06      55.000000       0.000000   
max    9.999816e+14   5.790484e+06     115.000000       1.000000   

        Hypertension       Diabetes     Alcoholism       Handicap  \
count  110527.000000  110527.000000  110527.000000  110527.000000   
mean        0.197246       0.071865       0.030400       0.022248   
std         0.397921       0.258265       0.171686       0.161543   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       4.000000   

         SMSReceived  
count  110527.000000  
mean        0.321026  
std         0.466873  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         1.000000  
max         1.000000

In [15]:
df.Age.describe()

count    110527.000000
mean         37.088874
std          23.110205
min          -1.000000
25%          18.000000
50%          37.000000
75%          55.000000
max         115.000000
Name: Age, dtype: float64

The minimum age in the dataset is -1 (which is not a valid age). To check it further:

In [16]:
df.query("Age == -1").shape[0]

1

There is only one row with age -1. 

In [17]:
df.query("Age == -1")

PatientId  AppointmentID Gender ScheduledDay AppointmentDay  Age  \
99832  465943158731293        5775010      F   2016-06-06     2016-06-06   -1   

      Neighbourhood  Scholarship  Hypertension  Diabetes  Alcoholism  \
99832         ROMÃO            0             0         0           0   

       Handicap  SMSReceived NoShow  
99832         0            0     No

Since Age cannot be negative, the row can be dropped using the index number

In [18]:
#df.drop(df.index[99832], inplace = True)

In [19]:
df.describe()

PatientId  AppointmentID            Age    Scholarship  \
count  1.105260e+05   1.105260e+05  110526.000000  110526.000000   
mean   1.474934e+14   5.675304e+06      37.089219       0.098266   
std    2.560943e+14   7.129544e+04      23.110026       0.297676   
min    3.921700e+04   5.030230e+06       0.000000       0.000000   
25%    4.172536e+12   5.640285e+06      18.000000       0.000000   
50%    3.173184e+13   5.680572e+06      37.000000       0.000000   
75%    9.438963e+13   5.725523e+06      55.000000       0.000000   
max    9.999816e+14   5.790484e+06     115.000000       1.000000   

        Hypertension       Diabetes     Alcoholism       Handicap  \
count  110526.000000  110526.000000  110526.000000  110526.000000   
mean        0.197248       0.071865       0.030400       0.022248   
std         0.397923       0.258266       0.171686       0.161543   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       4.000000   

         SMSReceived  
count  110526.000000  
mean        0.321029  
std         0.466874  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         1.000000  
max         1.000000

In [20]:
df.shape

(110526, 14)

Select the ages in ascending order

In [80]:
np.sort(df.Age.unique())

array([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,
        13,  14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,
        26,  27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,
        39,  40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,
        52,  53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,
        65,  66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,
        78,  79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,
        91,  92,  93,  94,  95,  96,  97,  98,  99, 100, 102, 115])

In [21]:
df.groupby('Age').count()['PatientId']

Age
0      3539
1      2273
2      1618
3      1513
4      1299
       ... 
98        6
99        1
100       4
102       2
115       5
Name: PatientId, Length: 103, dtype: int64

3539 of the patients are between 1 to 11 months old. Since they are infants, let check if they

Create a new column for the months for each scheduled appointment. This will help understand the number of scheduled appointments per month.

In [22]:
df['Month'] = pd.DatetimeIndex(df['ScheduledDay']).month

In [23]:

df.head(1)

PatientId  AppointmentID Gender ScheduledDay AppointmentDay  Age  \
0  29872499824296        5642903      F   2016-04-29     2016-04-29   62   

     Neighbourhood  Scholarship  Hypertension  Diabetes  Alcoholism  Handicap  \
0  JARDIM DA PENHA            0             1         0           0         0   

   SMSReceived NoShow  Month  
0            0     No      4

In [24]:
df.groupby('Month').count()

PatientId  AppointmentID  Gender  ScheduledDay  AppointmentDay    Age  \
Month                                                                          
1             60             60      60            60              60     60   
2            281            281     281           281             281    281   
3           3614           3614    3614          3614            3614   3614   
4          25339          25339   25339         25339           25339  25339   
5          67421          67421   67421         67421           67421  67421   
6          13749          13749   13749         13749           13749  13749   
11             1              1       1             1               1      1   
12            61             61      61            61              61     61   

       Neighbourhood  Scholarship  Hypertension  Diabetes  Alcoholism  \
Month                                                                   
1                 60           60            60        60          60   
2                281          281           281       281         281   
3               3614         3614          3614      3614        3614   
4              25339        25339         25339     25339       25339   
5              67421        67421         67421     67421       67421   
6              13749        13749         13749     13749       13749   
11                 1            1             1         1           1   
12                61           61            61        61          61   

       Handicap  SMSReceived  NoShow  
Month                                 
1            60           60      60  
2           281          281     281  
3          3614         3614    3614  
4         25339        25339   25339  
5         67421        67421   67421  
6         13749        13749   13749  
11            1            1       1  
12           61           61      61

In [25]:
df['ScheduledWeekDay'] = pd.to_datetime(df['ScheduledDay']).dt.day_name()

In [26]:
df.tail()

PatientId  AppointmentID Gender ScheduledDay AppointmentDay  \
110522    2572134369293        5651768      F   2016-05-03     2016-06-07   
110523    3596266328735        5650093      F   2016-05-03     2016-06-07   
110524   15576631729893        5630692      F   2016-04-27     2016-06-07   
110525   92134931435557        5630323      F   2016-04-27     2016-06-07   
110526  377511518121127        5629448      F   2016-04-27     2016-06-07   

        Age Neighbourhood  Scholarship  Hypertension  Diabetes  Alcoholism  \
110522   56   MARIA ORTIZ            0             0         0           0   
110523   51   MARIA ORTIZ            0             0         0           0   
110524   21   MARIA ORTIZ            0             0         0           0   
110525   38   MARIA ORTIZ            0             0         0           0   
110526   54   MARIA ORTIZ            0             0         0           0   

        Handicap  SMSReceived NoShow  Month ScheduledWeekDay  
110522         0            1     No      5          Tuesday  
110523         0            1     No      5          Tuesday  
110524         0            1     No      4        Wednesday  
110525         0            1     No      4        Wednesday  
110526         0            1     No      4        Wednesday

In [27]:
df.query("ScheduledDay == AppointmentDay").count()

PatientId           38562
AppointmentID       38562
Gender              38562
ScheduledDay        38562
AppointmentDay      38562
Age                 38562
Neighbourhood       38562
Scholarship         38562
Hypertension        38562
Diabetes            38562
Alcoholism          38562
Handicap            38562
SMSReceived         38562
NoShow              38562
Month               38562
ScheduledWeekDay    38562
dtype: int64

In [28]:
df.shape

(110526, 16)

In [29]:
df.head(1)

PatientId  AppointmentID Gender ScheduledDay AppointmentDay  Age  \
0  29872499824296        5642903      F   2016-04-29     2016-04-29   62   

     Neighbourhood  Scholarship  Hypertension  Diabetes  Alcoholism  Handicap  \
0  JARDIM DA PENHA            0             1         0           0         0   

   SMSReceived NoShow  Month ScheduledWeekDay  
0            0     No      4           Friday

Get the number of appointments based on Neighbourhood

In [30]:
df.groupby(['ScheduledWeekDay', 'NoShow']).count()['Age']

ScheduledWeekDay  NoShow
Friday            No        15028
                  Yes        3887
Monday            No        18523
                  Yes        4561
Saturday          No           23
                  Yes           1
Thursday          No        14373
                  Yes        3700
Tuesday           No        20877
                  Yes        5291
Wednesday         No        19383
                  Yes        4879
Name: Age, dtype: int64

In [31]:
df[['ScheduledWeekDay', 'NoShow']].value_counts()

AttributeError: 'DataFrame' object has no attribute 'value_counts'

In [35]:
df.Neighbourhood.value_counts()

JARDIM CAMBURI                 7717
MARIA ORTIZ                    5805
RESISTÊNCIA                    4431
JARDIM DA PENHA                3877
ITARARÉ                        3514
                               ... 
ILHA DO BOI                      35
ILHA DO FRADE                    10
AEROPORTO                         8
ILHAS OCEÂNICAS DE TRINDADE       2
PARQUE INDUSTRIAL                 1
Name: Neighbourhood, Length: 81, dtype: int64

Missing Appointment and Age

110526